### DISCUSSION TITLE
Slow initial page load with getServerSideProps

### DISCUSSION THREAD

**Discussion by denvudd** [vercel/next.js#50783] [Help] [answered]
### Summary

I am making a project about movies using Next.js. Everything works fine with regular page transitions, but when I try to navigate to the [id].tsx page that uses `getServerSideProps`, I experience a slow initial load of up to 7 seconds. In `getServerSideProps`, I fetch data from an API to display it in the components.

Initially, I used RTK Query with next-redux-wrapper in `getServerSideProps `to fetch the data, and then I accessed the data in the page component using the `useQuery ` hook. However, now I realize that I get the same result even if I use `fetch()`.

I also noticed that the slow initial load only occurs when deploying to Vercel or Netlify, not in development mode.

To be honest, no matter how much I search, I can't find the problem. Perhaps someone can help me identify the issue?

My GitHub repo: https://github.com/denvudd/react-dbmovies.github.io
My Vercel deployment: https://react-dbmovies.vercel.app/movies/popularMovies
Relative issues: 
1. [Super slow page load times in development environment](https://github.com/vercel/next.js/discussions/17977#top)
2. [First call to getServerSideProps is slow](https://github.com/vercel/next.js/discussions/12447#top)
3. [Long initial page load with dynamic routing and getServerSideProps (Issue in my repo)](https://github.com/denvudd/react-dbmovies.github.io/issues/2)

Results on initial load:
![image](https://github.com/vercel/next.js/assets/68691654/1f0993fc-e6b5-4c8a-84a4-a33a8f672872)

### Additional information

My next info:
```js
Operating System:
Platform: win32
Arch: x64
Version: Windows 10 Pro
 Binaries:
Node: 16.13.0
npm: N/A
Yarn: N/A
pnpm: N/A
 Relevant packages:
next: 13.4.2
eslint-config-next: N/A
react: 18.2.0
react-dom: 18.2.0
typescript: 5.0.4
```
My package.json:
```js
{
  "name": "react-tbmovies",
  "version": "0.1.0",
  "private": true,
  "scripts": {
    "dev": "next dev",
    "build": "next build",
    "start": "next start",
    "lint": "next lint"
  },
  "dependencies": {
    "@ant-design/icons": "^5.0.1",
    "@reduxjs/toolkit": "^1.9.5",
    "@types/node": "20.1.3",
    "@types/react": "18.2.6",
    "@types/react-dom": "18.2.4",
    "ahooks": "^3.7.7",
    "antd": "^5.4.7",
    "classnames": "^2.3.2",
    "clipboard": "^2.0.11",
    "dayjs": "^1.11.7",
    "lodash.debounce": "^4.0.8",
    "next": "13.4.2",
    "next-redux-wrapper": "^8.1.0",
    "numeral": "^2.0.6",
    "react": "18.2.0",
    "react-content-loader": "^6.2.1",
    "react-dom": "18.2.0",
    "react-redux": "^8.0.5",
    "react-youtube": "^10.1.0",
    "sass": "^1.62.1",
    "typescript": "5.0.4"
  },
  "devDependencies": {
    "@types/lodash.debounce": "^4.0.7",
    "@types/numeral": "^2.0.2"
  }
}
```
My [id].tsx:
```js
import React from "react";
import DetailLayout from "@/layouts/DetailsLayout";
import MovieDetailsBlock from "@/components/MovieDetailsBlock/MovieDetailsBlock";
import {
  getMovieDetails,
  getRunningQueriesThunk,
} from "@/redux/api/movies/slice";
import { wrapper } from "@/redux/store";
import { MovieDetails } from "@/redux/api/movies/types/MovieDetailsType";

interface MovieDetailsPageProps {
  id: number;
  data: MovieDetails;
}

export const getServerSideProps = wrapper.getServerSideProps(
  (store) => async (context) => {
    const { id } = context.query;

    if (typeof id === "string") {
      await store.dispatch(
        getMovieDetails.initiate({ id, params: "language=uk-UA&page=1" })
      );
    }

    const { data } = getMovieDetails.select({
      id,
      params: "language=uk-UA&page=1",
    })(store.getState());

    await Promise.all(store.dispatch(getRunningQueriesThunk()));

    return {
      props: { id, data },
    };
  }
);

const MovideDetailsPage: React.FC<MovieDetailsPageProps> = ({ id, data }) => {
  return (
    <>
      <DetailLayout>
        {data && <MovieDetailsBlock id={id} data={data} />}
      </DetailLayout>
    </>
  );
};

export default MovideDetailsPage;
```
My _app.tsx:
```js
import "@/styles/globals.scss";
import type { AppProps } from "next/app";
import AppLayout from "@/layouts/AppLayout";
import { wrapper } from "@/redux/store";
import { Provider } from "react-redux";

const App = ({ Component, ...rest }: AppProps) => {
  const { store, props } = wrapper.useWrappedStore(rest);
  return (
    <Provider store={store}>
      <AppLayout>
        <Component {...props.pageProps} />
      </AppLayout>
    </Provider>
  );
}

export default App;
```

**Comment by denvudd** [comment]
This is an example of data that fetches from the API:
```js
{
  "adult": false,
  "backdrop_path": "/hZkgoQYus5vegHoetLkCJzb17zJ.jpg",
  "belongs_to_collection": null,
  "budget": 63000000,
  "genres": [
    {
      "id": 18,
      "name": "Drama"
    },
    {
      "id": 53,
      "name": "Thriller"
    },
    {
      "id": 35,
      "name": "Comedy"
    }
  ],
  "homepage": "http://www.foxmovies.com/movies/fight-club",
  "id": 550,
  "imdb_id": "tt0137523",
  "original_language": "en",
  "original_title": "Fight Club",
  "overview": "A ticking-time-bomb insomniac and a slippery soap salesman channel primal male aggression into a shocking new form of therapy. Their concept catches on, with underground \"fight clubs\" forming in every town, until an eccentric gets in the way and ignites an out-of-control spiral toward oblivion.",
  "popularity": 61.416,
  "poster_path": "/pB8BM7pdSp6B6Ih7QZ4DrQ3PmJK.jpg",
  "production_companies": [
    {
      "id": 508,
      "logo_path": "/7cxRWzi4LsVm4Utfpr1hfARNurT.png",
      "name": "Regency Enterprises",
      "origin_country": "US"
    },
    {
      "id": 711,
      "logo_path": "/tEiIH5QesdheJmDAqQwvtN60727.png",
      "name": "Fox 2000 Pictures",
      "origin_country": "US"
    },
    {
      "id": 20555,
      "logo_path": "/hD8yEGUBlHOcfHYbujp71vD8gZp.png",
      "name": "Taurus Film",
      "origin_country": "DE"
    },
    {
      "id": 54051,
      "logo_path": null,
      "name": "Atman Entertainment",
      "origin_country": ""
    },
    {
      "id": 54052,
      "logo_path": null,
      "name": "Knickerbocker Films",
      "origin_country": "US"
    },
    {
      "id": 4700,
      "logo_path": "/A32wmjrs9Psf4zw0uaixF0GXfxq.png",
      "name": "The Linson Company",
      "origin_country": "US"
    },
    {
      "id": 25,
      "logo_path": "/qZCc1lty5FzX30aOCVRBLzaVmcp.png",
      "name": "20th Century Fox",
      "origin_country": "US"
    }
  ],
  "production_countries": [
    {
      "iso_3166_1": "US",
      "name": "United States of America"
    }
  ],
  "release_date": "1999-10-15",
  "revenue": 100853753,
  "runtime": 139,
  "spoken_languages": [
    {
      "english_name": "English",
      "iso_639_1": "en",
      "name": "English"
    }
  ],
  "status": "Released",
  "tagline": "Mischief. Mayhem. Soap.",
  "title": "Fight Club",
  "video": false,
  "vote_average": 8.433,
  "vote_count": 26280
}
```

**Comment by icyJoseph** [comment]
I wanted to run a test on your `[id].tsx` page, but it looks like you need to be authenticated/authorized for that?

## Edit

It worked haha, it was just slow... alright, let me see what I can do

**Comment by icyJoseph** [comment]
@denvudd I suspect the problem here are some of the supporting libraries, for example `antd`. More precisely, they might not be properly tree-shaken, which would naturally lead to the entire library being parsed, per request. You don't see this in SPA transitions, because as soon as a link pointing to `[id].tsx` is on screen Next.js fetches the JS to drive that page, so while the user scrolls around the JS for that page is generated and downloaded. 

However, upon first navigation, the entire HTML ought to be generated, and those `import { Something } from 'library'` look very scary. The parser will go into "library" and follow every import statement, and so on parsing all of that. It is the best lead we have, so maybe change to `import Something from 'library/something'`, wherever possible? See if that helps out. Not only for component libraries, but utils, etc...

**Reply by denvudd** [reply]
Hello, thank you for your response!

That's strange because the [Ant Design documentation states](https://ant.design/docs/react/getting-started#import-on-demand) that the latest versions of the library support tree shaking:
> Import on Demand
> antd supports tree shaking of ES modules, so using import { Button } from 'antd'; would drop js code you didn't use.

I tried removing all the components from Ant Design in `[id].tsx` and leaving only the example from the Next.js documentation that uses regular fetch and passes data. However, even in that case, the loading takes about 7 seconds. 
To be honest, I am completely confused by this...
```js
import React from "react";
import { MovieDetails } from "@/redux/api/movies/types/MovieDetailsType";

import { GetServerSideProps } from "next";

interface MovieDetailsPageProps {
  id: number;
  data: MovieDetails;
}

export const getServerSideProps: GetServerSideProps = async () => {
  const res = await fetch('https://api.github.com/repos/vercel/next.js');
  const repo = await res.json();
  return { props: { data: repo } };
};

const MovideDetailsPage: React.FC<MovieDetailsPageProps> = ({ data }) => {
  
  return (
    <>
      <div style={{ color: "#000" }}>{data?.id}</div>
    </>
  );
};

export default MovideDetailsPage;
```


I also tried tree shaking in `@ant-design/icons` (just in case), but it didn't help either. Also, in one of the issues, [there was a suggestion](https://github.com/vercel/next.js/discussions/17977#discussioncomment-6084887) to use `modularizeImports` (there was a case with Material UI). To be honest, I'm still learning Next.js and I don't quite understand what it is and how it can help me. I'll try it in the next few days and write here about any changes, but for now, I have very few leads and guesses...

**Reply by icyJoseph** [reply]
I can help a bit more tomorrow mate.

The thing is that in a situation like this, making assumptions is just not gonna help. We must put a question mark on everything and start discarding.

In hindsight, it could have been better to use `@next/bundle-analyzer` and see if there are any large chunks for the given entry point.

I'd recommend setting it up anyway.

I'll also start to place some performance measurements on the GSSP function. Read up on node performance marks and measurements.

**Reply by denvudd** [reply]
I analyzed the bundle using `@next/bundle-analyzer`, thanks. It's my first time using it, and I don't fully understand it, but in analyze/nodejs, I can see that ` ../pages/movie/[id].js ` weighs quite a lot. Honestly, I'm not entirely sure if this is normal or not.

<details>
  <summary>Full analyze/nodejs:</summary>
  <img src="https://github.com/vercel/next.js/assets/68691654/c68683a5-a15b-477a-8a90-0ad8fd4006d7">
</details>

<details>
  <summary>Only [id].js:</summary>
  <img src="https://github.com/vercel/next.js/assets/68691654/9a49b336-4f2e-476d-9d92-9b9f855f3652">
</details>

<details>
  <summary>Tree map sizes:</summary>
  <img src="https://github.com/vercel/next.js/assets/68691654/50b2fc5d-66e8-4e45-80cf-83718ccb7aad">
</details>

<details>
  <summary>Also full analyze/client:</summary>
  <img src="https://github.com/vercel/next.js/assets/68691654/51688872-0ade-4e85-8dfc-142e2a1d5d81">
</details>

<details>
  <summary>And only pages/movies/[id]:</summary>
  <img src="https://github.com/vercel/next.js/assets/68691654/1521bea3-0f8d-4a37-a74d-732dc79176bd">
</details>

Also,  I tried not using `css-in-js` (just in case) in `_document.tsx` (I used it because of the[ "Flash of unstyled content for first load" issue](https://github.com/ant-design/ant-design/issues/16037#issuecomment-483140458 )), but the results didn't change.

**Reply by icyJoseph** [reply]
I wonder if the issue is just that the page is heavy and the cold start of your page is too large.

If you make a `/pages/raw/movies/[id]` route and just return the `JSON.stringify(data)`, like, no components, just raw JSON. Is that page slow?

**Reply by denvudd** [reply]
I created `pages/raw/movies/[id].tsx` file with the code from next.js documentation... and it also takes 7 seconds to load on Vercel. In Dev mode, everything still loads at normal speed (600 milliseconds). From the antd library, the only component used in `pages/raw/movies/[id].tsx` is AppLayout, which utilizes the Layout component and wraps the entire application. However, I don't believe the issue lies with it (I checked). Now I definitely don't understand anything haha
Could the issue be with Vercel/Netlify themselves? Maybe in settings or idk :(

`pages/raw/movies/[id].tsx`:
```js
import type { InferGetServerSidePropsType, GetServerSideProps } from 'next';
 
type Repo = {
  name: string;
  stargazers_count: number;
};
 
export const getServerSideProps: GetServerSideProps<{
  repo: Repo;
}> = async () => {
  const res = await fetch('https://api.github.com/repos/vercel/next.js');
  const repo = await res.json();
  return { props: { repo } };
};
 
export default function Page({
  repo,
}: InferGetServerSidePropsType<typeof getServerSideProps>) {
  return repo.stargazers_count;
}
```

The `MovieCard` component contains a link to navigate:
```js
import React from "react";
import { Card, Typography } from "antd";
import Image from "next/image";
import Link from "next/link";
import RatingBar from "../RatingBar/RatingBar";
import { formatReleaseDate } from "@/utils/formatReleaseDate";

import styles from "./MovieCard.module.scss";

interface MovieCardProps {
  id: number;
  index: number;
  title: string;
  imgUrl: string;
  description: string;
  voteAverage: number;
  release: string;
}

const MovieCard: React.FC<MovieCardProps> = ({
  id,
  index,
  title,
  imgUrl,
  description,
  voteAverage,
  release,
}) => {
  const { Title, Paragraph } = Typography;

  return (
    <Card
      hoverable
      size="small"
      cover={
        <Link href={`/raw/movies/${id}`}>
          <div className={styles.containerImage}>
            <Image
              className={styles.cardImage}
              alt="alt"
              width={300}
              height={400}
              src={imgUrl}
              priority={index < 10} // first 10 images priority = true, the rest will not
            />
          </div>
        </Link>
      }
    >
      <p>{formatReleaseDate(release)}</p>
      <Link href={`/raw/movies/${id}`}>
        <Title ellipsis={{ rows: 1 }} level={5}>
          {title}
        </Title>
      </Link>
      <Paragraph ellipsis={{ rows: 2 }}>{description}</Paragraph>
      <div className={styles.cardRating}>
        <RatingBar rating={voteAverage} size={40}/>
      </div>
    </Card>
  );
};

export default MovieCard;
```
JSON id:
![image](https://github.com/vercel/next.js/assets/68691654/6b6ba1d3-9793-4963-9f91-5f4d7e16fbc1)

Preview:
![image](https://github.com/vercel/next.js/assets/68691654/32fa1640-052c-4695-9898-f014a9b3e8fd)

Thanks for your answers anyway

**Comment by denvudd** [comment]
So, here are the things I've tried:
1. Creating a new dynamic page (`pages/raw/movies/[id].tsx`) and using a regular `fetch()` in `getServerSideProps`, as described in the [Next.js documentation](https://nextjs.org/docs/pages/building-your-application/data-fetching/get-server-side-props) - **doesn't work.**
2. Not using ANTD components on the page (even though the documentation states that ANTD [supports tree shaking](https://ant.design/docs/react/getting-started#import-on-demand)) - **doesn't work.**
3. Changing the Function Region in Settings on Vercel. I switched from the default value of Washington, USA to Frankfurt, Germany (I live in Ukraine) - **doesn't work.**
4. Trying Netlify instead of Vercel - **doesn't work.**
5. Not using ANTD's `css-in-js` (I used it because of the "[Flash of unstyled content for first load](https://github.com/ant-design/ant-design/issues/16037#issuecomment-483140458 )" issue) - **doesn't work.**


The current behavior of the page is as follows:

Clicking on a link triggers a request for a JSON with the movie ID **->** On the first load, it takes up to 7 seconds to load the JSON **->** The requested data is displayed along with HTML and CSS markup **->** Subsequent navigations happen quickly, but if you wait for approximately a minute (or maybe longer), the page reverts to the state it was in after the initial click, and the cycle repeats.

In development mode, the loading occurs normally.

Honestly, I'm exhausted from searching for the problem... Either I'm missing something super obvious, or I don't understand something super complicated :(

**Reply by icyJoseph** [reply]
If you make an API route under pages, that just replies with hello world, like:

```js
export default function handler(req, res) {
  res.status(200).json({ name: 'John Doe' })
}
```

Does it also take a long time to respond? 

Do you have other GetServerSideProps in the app? Are those slow too?

I wonder if this is the problem: https://vercel.com/docs/concepts/functions/serverless-functions#bundling-serverless-functions

It would be like this, your app bundle, is too large, and the lambda that responds to GetServerSideProps is part of that app bundle. You could try to separate it as they show there, and see if that helps.

**Reply by denvudd** [reply]
I have created an API endpoint `/api/hello` that outputs the handler function with JSON. The first load takes up to ~2 seconds, and then the behavior repeats, similar to `getServerSideProps`. Well, at least it's no longer seven seconds... To answer your question, yes, I have another `getServerSideProps` that currently only retrieves the id from `context.query`, and it's also slow, just like in `[id].tsx` in the `pages/movies` directory.

I will try to investigate what you shared. Thank you. However, at first glance, I'm not entirely sure how to do it. Do I need to send requests through API Routes and "split" them in the function configuration? Well, I'll try to figure it out and test it.

### Edit
Also, if the issue lies with a heavy bundle during loading, honestly, I don't know how to make it lighter. Even when the page is empty without any imports, it still produces the same result. If the problem is with the Ant Design library, I have no idea how to make it lighter either. If I try to extract each component like `antd/src/dist/modal`, it often throws an error like `"Cannot use import statement outside a module"` or something like this. Moreover, the Ant Design documentation claims to support tree-shaking, so a simple `import {something} from 'antd'` should work, right? Would using dynamic imports, for example, help solve the problem? It seems like I'm missing something.

**Comment by denvudd** [comment]
Also, I tried creating a new project from scratch just to clarify everything. I installed only Next.js and the ANTD library, created `pages/movies/[id].tsx` with an example from the Next.js documentation, and deployed it on Vercel. The cold start takes about 2 seconds to load, while the warm start takes around 100 milliseconds.

Links:
Vercel Deploy - https://tmdb-movies-test-82lr.vercel.app/
GitHub Repo - https://github.com/denvudd/tmdb-movies-test

package.json:
```js
{
  "name": "tbmovies-test",
  "version": "0.1.0",
  "private": true,
  "scripts": {
    "dev": "next dev",
    "build": "next build",
    "start": "next start",
    "lint": "next lint"
  },
  "dependencies": {
    "@types/node": "20.3.1",
    "@types/react": "18.2.12",
    "@types/react-dom": "18.2.5",
    "antd": "^5.6.1",
    "next": "13.4.5",
    "react": "18.2.0",
    "react-dom": "18.2.0",
    "typescript": "5.1.3"
  }
}
```

[id].tsx:
```js
import type { InferGetServerSidePropsType, GetServerSideProps } from 'next'
 
type Repo = {
  name: string
  stargazers_count: number
}
 
export const getServerSideProps: GetServerSideProps<{
  repo: Repo
}> = async () => {
  const res = await fetch('https://api.github.com/repos/vercel/next.js')
  const repo = await res.json()
  return { props: { repo } }
}
 
export default function Page({
  repo,
}: InferGetServerSidePropsType<typeof getServerSideProps>) {
  return repo.stargazers_count
}
```

Cold start results:
![image](https://github.com/vercel/next.js/assets/68691654/aa670ddf-eed1-4e22-ab08-42e892d912bd)

Bundle:
![image](https://github.com/vercel/next.js/assets/68691654/239cbef6-e08d-4647-81b2-78916f92f3b2)

**Reply by denvudd** [reply]
@icyJoseph Thanks, mate for pointing me in the right direction to troubleshoot the issue and show me something new :)

**Comment by denvudd** [comment, marked as answer]
I have resolved this issue by adding the following code to `pages/movies/[id].tsx:`
```js
export const config = {
  runtime: 'experimental-edge', // warn: using an experimental edge runtime, the API might change
}
```
Results of cold start:
![image](https://github.com/vercel/next.js/assets/68691654/74769a1a-837d-4f7d-b1d5-b6aa69522aba)

Now, the cold start times are around 200-300 milliseconds. I still don't know why GSSP and API Routes take a long time to load during cold starts even in the new Next.js project from scratch. I have only found that I am not alone in experiencing this issue, and the earliest mentions of this problem date back to 2020. As far as I understand, Vercel has not commented on this problem.
I'm not aware of the downsides and potential issues of using this solution, but for my purposes, it is suitable. I would appreciate it if someone could point out any potential issues with using this method.

To read more about it: https://github.com/denvudd/react-dbmovies.github.io/issues/2#issuecomment-1591612339

**Reply by MauroSerrano-dev** [reply]
I have to test a little more, but I think this solved my problem. Thank you very much <3

**Reply by denvudd** [reply]
I am glad to hear that!

**Reply by MauroSerrano-dev** [reply]
After some tests I realized that in production it works normally, the page opens quickly. However, in dev using the runtime: 'experimental-edge' causes some conflict with @mui/material and the app crashes

**Reply by TomasSestak** [reply]
Cant be used for all the APIs node has

**Reply by liho00** [reply]
this is only applicable on vercel right? how about self deployment?

**Comment by TomasSestak** [comment]
makes SSR unusable - isn't this the main reason for Next?
